pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,464 @@
+    gradient.
+
+    By keeping a running average of the parameter-shift gradient and the *variance*
+    of the parameter-shift gradient, this optimizer frugally distributes a shot
+    budget across the partial derivatives of each parameter.
+
+    In addition, if computing the expectation value of a Hamiltonian using
+    :class:`~.ExpvalCost`, weighted random sampling can be used to further
+    distribute the shot budget across the local terms from which the Hamiltonian
+    is constructed.
+
+    .. note::
+
+        The shot adaptive optimizer only supports single QNodes or :class:`~.ExpvalCost` objects as
+        objective functions. The bound device must also be instantiated with a finite number
+        of shots.
+
+    Args:
+        min_shots (int): The minimum number of shots used to estimate the expectations
+            of each term in the Hamiltonian. Note that this must be larger than 2 for the variance
+            of the gradients to be computed.
+        mu (float): The running average constant :math:`\mu \in [0, 1]`. Used to control how quickly the
+            number of shots recommended for each gradient component changes.
+        b (float): Regularization bias. The bias should be kept small, but non-zero.
+        weighted_random_sampling (bool): Whether to use the weighted random sampling algorithm
+            to multinomially distribute the shot budget across terms in the Hamiltonian expectation value.
+            Only takes effect if the objective function provided is an instance of :class:`~.ExpvalCost`.
+        stepsize (float): The learning rate :math:`\eta`. The learning rate *must* be such
+            that :math:`\eta < 2/L = 2/\sum_i|c_i|`, where:
+
+            * :math:`L \leq \sum_i|c_i|` is the bound on the `Lipschitz constant
+              <https://en.wikipedia.org/wiki/Lipschitz_continuity>`__ of the variational quantum
+              algorithm objective function, and
+
+            * :math:`c_i` are the coefficients of the Hamiltonian used in the objective function.
+
+    The shot adaptive optimizer is based on the iCANS1 optimizer by
+    `Kübler et al. (2020) <https://quantum-journal.org/papers/q-2020-05-11-263/>`__, and works
+    as follows:
+
+    1. The initial step of the optimizer is performed with some specified minimum
+       number of shots, :math:`s_{min}`, for all partial derivatives.
+
+    2. The parameter-shift rule is then used to estimate the gradient :math:`g_i`
+       for each parameter :math:`\theta_i`, parameters, as well as the *variances*
+       :math:`v_i` of the estimated gradients.
+
+    3. Gradient descent is performed for each parameter :math:`\theta_i`, using
+       the pre-defined learning rate :math:`\eta` and the gradient information :math:`g_i`:
+       :math:`\theta_i \rightarrow \theta_i - \eta g_i`.
+
+    4. The improvement in the expected gain per shot, for a specific parameter value,

```

Comment:
Something along the lines of "A maximum shot number is set where this value is maximized".  Right now it seems like this is calculated and then is just a handy but useless piece of information.

Reply:
```suggestion
    4. A maximum shot number is set by maximizing the improvement in the expected gain per shot.
       For a specific parameter value, the improvement in the expected gain per shot
```